lib/compress/zstd_lazy.c

```diff
@@ -531,8 +569,74 @@ size_t ZSTD_HcFindBestMatch_generic (
         matchIndex = NEXT_IN_CHAIN(matchIndex, chainMask);
     }
 
-    if (dictMode == ZSTD_dictMatchState) {
-        const ZSTD_matchState_t* const dms = ms->dictMatchState;
+    if (dictMode == ZSTD_dedicatedDictSearch) {
+        const U32 ddsChainSize    = (1 << dms->cParams.chainLog);
+        const U32 ddsChainMask    = ddsChainSize - 1;
+        const U32 ddsLowestIndex  = dms->window.dictLimit;
+        const BYTE* const ddsBase = dms->window.base;
+        const BYTE* const ddsEnd  = dms->window.nextSrc;
+        const U32 ddsSize         = (U32)(ddsEnd - ddsBase);
+        const U32 ddsIndexDelta   = dictLimit - ddsSize;
+        const U32 ddsMinChain     = ddsSize > ddsChainSize ? ddsSize - ddsChainSize : 0;
+        const U32 bucketSize      = (1 << ZSTD_LAZY_DDSS_BUCKET_LOG);
+        const U32 bucketLimit     = nbAttempts < bucketSize ? nbAttempts : bucketSize;
+        U32 ddsAttempt;
+
+        for (ddsAttempt = 0; ddsAttempt < bucketSize; ddsAttempt++) {
```

Comment:
I see 20% of this functions cycles spent in this loop. That suggests to me that the earlier prefetch isn't early enough. Can we do better?

Reply:
Even with the prefetch ip+1 change I suggested below we are spending 10% of cycles in this loop. So we could likely do even better.

Also prefetching ip+2 gives another 2% of speed in cold dictionaries. Fetching ip+2 we still spend 8% of cycles in this loop. I don't think that simple strategy is the right approach, but it shows that we can to do better in the final version of the code.

For example, we could move the prefetch out of the search function. Then:
* If we did not find a match prefetch the next (1, 2, 3, ...?) positions? Maybe # of positions depends on strategy. Would have to take care not to double-prefetch. This makes sense for lazy because we will search the next few positions, even if we find a match in the first search, since this is the lazy strategy. In greedy, it still makes sense, because we may predict we will find a few literals before a match. But maybe we should be less aggressive.
* If we find a match prefetch ip + matchLength + (0, 1, or 2, ...) positions before doing anything else. Similar to above. But we should also consider rep-codes, and we may be more likely to find a 2nd match directly after a match than searching a position where we just saw a literal.